HiP-DataStore/Core/ReadModel/CacheDatabaseManager.cs

```diff
@@ -0,0 +1,230 @@
+using System;
+using Tag = PaderbornUniversity.SILab.Hip.DataStore.Model.Entity.Tag;
+
+namespace PaderbornUniversity.SILab.Hip.DataStore.Core.ReadModel
+{
+    /// <summary>
+    /// Subscribes to EventStore events to keep the cache database up to date.
+    /// </summary>
+    public class CacheDatabaseManager
+    {
+        private readonly EventStoreClient _eventStore;
+        private readonly IMongoDatabase _db;
+        private readonly ILogger<CacheDatabaseManager> _logger;
+
+        public IMongoDatabase Database => _db;
+
+        public CacheDatabaseManager(
+            EventStoreClient eventStore,
+            IOptions<EndpointConfig> config,
+            ILogger<CacheDatabaseManager> logger)
+        {
+            // For now, the cache database is always created from scratch by replaying all events.
+            // This also implies that, for now, the cache database always contains the entire data (not a subset).
+            // In order to receive all the events, a Catch-Up Subscription is created.
+
+            _logger = logger;
+
+            // 1) Open MongoDB connection and clear existing database
+            var mongo = new MongoClient(config.Value.MongoDbHost);
+            mongo.DropDatabase(config.Value.MongoDbName);
+            _db = mongo.GetDatabase(config.Value.MongoDbName);
+            logger.LogInformation($"Connected to MongoDB cache database, using database '{config.Value.MongoDbName}'");
+
+            // 2) Subscribe to EventStore to receive all past and future events
+            _eventStore = eventStore;
+
+            _eventStore.Connection.SubscribeToStreamFrom(
+                config.Value.EventStoreStream,
+                null, // don't use StreamPosition.Start (see https://groups.google.com/forum/#!topic/event-store/8tpXJMNEMqI),
+                CatchUpSubscriptionSettings.Default,
+                OnEventAppeared);
+        }
+
+        private void OnEventAppeared(EventStoreCatchUpSubscription subscription, ResolvedEvent resolvedEvent)
+        {
+            try
+            {
+                // Note regarding migration:
+                // Event types may change over time (properties get added/removed etc.)
+                // Whenever an event has multiple versions, an event of an obsolete type should be transformed to an event
+                // of the latest version, so that ApplyEvent(...) only has to deal with events of the current version.
+
+                var ev = resolvedEvent.Event.ToIEvent().MigrateToLatestVersion();
+                ApplyEvent(ev);
+            }
+            catch (Exception e)
+            {
+                _logger.LogWarning($"{nameof(CacheDatabaseManager)} could not process an event: {e}");
+            }
+        }
+        
+        private void ApplyEvent(IEvent ev)
+        {
+            switch (ev)
+            {
+                case ExhibitCreated e:
+                    var newExhibit = new Exhibit(e.Properties)
```

Comment:
Fair point, but in this particular case this is the result of a C# annoyance: In C#, the scope of a variable declared in a `case` spans the whole `switch`, so I can't declare the same variables in multiple cases. And instead of `exhibit`, `exhibit2`, `exhibit3` etc., I prefer `newExhibit`, `updatedExhibit` and the like.